### ISSUE TITLE
AlexNet finetune on Office-31

### ISSUE THREAD

**Issue by qinenergy** [jindongwang/transferlearning#23]
Hey, I am having the exact same problem on reproducing the Office-31 finetune results. I am currently trying to reproduce the 64.2% source-only baseline on Amazon->Webcam task in this paper:
https://arxiv.org/pdf/1602.03534.pdf

So far, the best result I get is around 56% using the following setup, which is higher than your reported 46%.  Something is still wrong tho.

Network:       .... fc7 -> fc(128) ->fc(31)
lr:                  2.5e-3
Preprocess:  Resize(256) + RandomCrop(227)
Optimizer:     tf.train.MomentumOptimizer(lr, 0.9)
Trainable:      Only fc6, fc7, fc(128), fc(31)
No weight decay.

Let's keep in contact QwQ

**Comment by jindongwang** [comment]
@qinenergy Thank you for your reply! I will try your parameter setting to see what results I get. Keep in touch.

**Comment by jindongwang** [comment]
@qinenergy Could you please explain to me why your network has 9 layers? AlexNet has 8 layers with the final layers as fc6, fc7, and fc8.

**Comment by wogong** [comment]
@jindongwang IMHO, fc(128) is bottleneck layer, many DA papers using bottleneck layer in finetune.

**Comment by jindongwang** [comment]
@wogong But it is only the results of train on source, and test on target. There is no adaptation. Why bottleneck?

**Comment by wogong** [comment]
@jindongwang 

emmm, my mistake. I thought this is talking about adaptation. I agree that there is no need for bottleneck layer in source-only baseline.

**Comment by qinenergy** [comment]
@jindongwang Hey, I agree with you that it is unnecessary to have a bottleneck layer in the source only case.

**Comment by jindongwang** [comment]
@qinenergy @wogong Then it's mysterious for the Office dataset. We can only consider using other popular datasets. This dataset is so awkward.

**Comment by wogong** [comment]
@jindongwang Have you tried Long's Xlearn repo <https://github.com/thuml/Xlearn> for reproducing the alexnet finetune results?
They use very different data pre-process code, like custom defined crop in the PyTorch part.
I am working on domain adaptation using PyTorch right now and come across the same problem as you. Nearly all current algorithms cannot get claimed accuracy on Office dataset using PyTorch implementations.

**Comment by jindongwang** [comment]
@wogong Long's data processing step can only help a little, which can be ignored.
It looks like no one can successfully reimplement the results of AlexNet and ResNet using Pytorch.

**Comment by jindongwang** [comment]
@wogong That's much better than what I got. Could you please modify the code in this repo to replace with your parameter settings?

**Comment by wogong** [comment]
@jindongwang I tried but failed to get the code run correctly. 😂
You can try my released repo https://github.com/wogong/pytorch-office_finetune.

**Comment by wogong** [comment]
BTW, I found this link in PyTorch forum. https://discuss.pytorch.org/t/pytorch-alexnet-not-as-good-as-original-alexnet-implementation/16680

Maybe the problem lies on torchvision's pretrained AlexNet model.

**Comment by jindongwang** [comment]
@wogong I also couldn't run your code successfully...
I know the LRN thing in AlexNet, but it seems doesn't help. A lot of people have demonstrated that LRN is not helping.
I tried your resize 227, it really works. Now my accuracy is above 50%.
I do not understand your code: why are there train-eval and test-eval dataset in loading the data?
And, have you tried ResNet?

**Comment by wogong** [comment]
@jindongwang emmm, I have not clear my code, so it may be confusing. src_eval and tgt_train is unnecessary. I'll try ResNet later.

**Comment by jindongwang** [comment]
@wogong @qinenergy I think we are making mistakes about the results from other papers. The evaluation **protocol** is different.
According to existing work (DDC, DAN, JAN, RTN, DCORAL etc.), the accuracy of AlexNet and ResNet are 61% and 80%. However, they are using different protocols: they take 20 samples per class (or 8 for dslr) as the source and target domains (we call it **down-sample**). Ours use the whole domain data (we call it **full-training**). This is quite different.
Actually, almost all the deep transfer learning methods are following this down-sampled protocol. Some research used the **full-training** protocol, but they just copy the results from the **down-sample** papers. Finally, they confuse us.

**Comment by jindongwang** [comment]
@wogong @qinenergy I checked the results again. They are using the full-training protocol. Still, we can't reproduce the results. Maybe it's time to give up Office.

**Comment by qinenergy** [comment]
Hey, I managed to get ~61% accuracy using Tensorflow on the A->W task, which I believe is good enough. I wrote a [blog post](https://www.qin.ee/2018/06/25/da-office/) about the tricks 

Here are a few things I noticed that might effect performance:

**Conv paddings**: They are defined differently in TF and Caffe which can hurt performance when loading pretrained caffe weights.
**lrn layer**: Since I use the caffe pretrained weights, I have to use the lrn layers with same parameters.
**Data preprocessing**: cv2.resize(256) -> randomcrop(227) I found that following the eact pipeline boosts the performance. I mean exact, use cv2, even tf.image could hurt performance.
**Lr setting**: I am not sure whether this helps, but I followed the eact lr settings of the GRL paper, which means 1x for pretrained kernels, 2x for pretrained biases, 10x for fc(31), 20x for fc(31) bias.

I don’t use PyTorch, so I am not sure the problems I encountered are the same as yours. Hope this helps.
